src/stores/index.js

```diff
@@ -929,40 +834,120 @@ const Store = Reflux.createStore({
    *
    * @param {Object} connection - The current connection.
    */
-  _connect(connection) {
+  async _connect(connection) {
+    // Set the connection's app name to the electron app name of Compass.
+    connection.appname = electron.remote.app.getName();
+
     this.dataService = new DataService(connection);
     this.appRegistry.emit('data-service-initialized', this.dataService);
-    this.dataService.connect((error, ds) => {
-      if (error) {
-        this.StatusActions.done();
-        this.setState({
-          isValid: false,
-          errorMessage: error.message,
-          syntaxErrorMessage: null
-        });
-      } else {
-        const currentConnection = this.state.currentConnection;
-        const currentSaved = this.state.connections[currentConnection._id];
 
-        this.state.isValid = true;
-        this.state.isConnected = true;
-        this.state.errorMessage = null;
-        this.state.syntaxErrorMessage = null;
-        this.state.hasUnsavedChanges = false;
-        this.state.isURIEditable = false;
-        this.state.customUrl = this.state.currentConnection.driverUrl;
+    try {
+      const runConnect = promisify(this.dataService.connect.bind(this.dataService));
+      const connectedDataService = await runConnect();
 
-        currentConnection.lastUsed = new Date();
+      const currentConnection = this.state.currentConnection;
+      const currentSaved = this.state.connections[currentConnection._id];
 
-        if (currentSaved) {
-          this._saveConnection(currentConnection);
-        } else {
-          this._saveRecent(currentConnection);
-        }
+      this.state.isValid = true;
```

Comment:
is it intentional that we mutate the state in the same object? what about just:

```
this.state = { ... this.state, isValid: true, ...etc }
```

Reply:
also why we don't do setState here?

Reply:
There are two types of actions to change reflux state:

```
const isValid = this.state.isValid;
this.trigger(this.state);
```

```
this.setState({
  isValid: false,
});
```

They drove me crazy because I didn't get why sometimes only one of those works for the plugin and tests. I hope we refactor the connection model soon and get rid of the ampersand model and then from reflux in the compass connct.